### DISCUSSION TITLE
Usage of `defaultItems` on virtualized combobox grid

### DISCUSSION THREAD

**Discussion by sclebal** [ariakit/ariakit#4240] [Questions] [answered]
👋🏻 I'm experimenting with the implementation of a virtualized combobox grid using [@tanstack-virtual](https://tanstack.com/virtual/latest). Here's a CodeSandbox with the implementation https://codesandbox.io/p/sandbox/virtualized-combobox-with-tanstack-virtual-dy39dk. In the process of implementing this, I first implemented a list and later the grid. Though keyboard navigation was perfectly working on the list, when I started using `rowId` the arrow navigation began to fail. To reproduce the bug you have to remove the `defaultItems` prop on the CodeSandbox and do:

1. _Considering the focus is on the input..._
2. Press "ArrowUp": Item 999 will be focused.
3. Press "ArrowDown": focus will be on the base element.
4. Press "ArrowUp": Item 996 will be focused instead of Item 999.
5. Press "ArrowDown" couple times: focus will be on the base element.
6. Press "ArrowDown": Item 0 will be focused.
7. Press "ArrowUp": focus will be on the base element
8. Press "ArrowUp": Item 36 (may be a different one based on your screen size) will be focused instead of Item 999.

I was inspired to use the `defaultItems` prop based on this @diegohaz's example [select-combobox-virtualized](https://ariakit.org/examples/select-combobox-virtualized) and it worked.

Trying to understand why I had to use `defaultItems` I spotted this line https://github.com/ariakit/ariakit/blob/91e7f3d8caf356e38c8fbe5e0d5ff29e53c1d030/packages/ariakit-react-core/src/composite/composite.tsx#L391 on the `Composite` code. I wondered if using `items` instead of `renderedItems` was intended or it's a bug. Mainly because AFAIK `renderedItems` is DOM-sorted but `items` is based on `registerItem` call order, so on cases like virtualized lists it won't be ordered and `up` fn should consider items ordering.

I open this discussion with the aim of understanding better Ariakit Composite grid and helping anyone who's struggling to implement this sort of virtualized combobox.

**Comment by diegohaz** [comment]
> Trying to understand why I had to use `defaultItems` I spotted this line
> 
> https://github.com/ariakit/ariakit/blob/91e7f3d8caf356e38c8fbe5e0d5ff29e53c1d030/packages/ariakit-react-core/src/composite/composite.tsx#L391
> 
> on the `Composite` code. I wondered if using `items` instead of `renderedItems` was intended or it's a bug. Mainly because AFAIK `renderedItems` is DOM-sorted but `items` is based on `registerItem` call order, so on cases like virtualized lists it won't be ordered and `up` fn should consider items ordering.

I believe that might be a bug, likely a remnant from when `items` were ordered like `renderedItems`.

**Reply by diegohaz** [reply, marked as answer]
This should be fixed now.
